src/service-worker.js

```diff
@@ -0,0 +1,421 @@
+ * sent to clients with timeouts from the rotation request in the attribute
+ * 'tokenExpiration'. The response is a resolved Promise.
+ *
+ * When RTR fails, a new message with type === RTR_ERROR is sent to clients
+ * with additional details in the attribute 'error'. The response is a
+ * rejected Promise.
+ *
+ */
+
+
+/** { atExpires, rtExpires } both are JS millisecond timestamps */
+let tokenExpiration = null;
+
+/** string FQDN including protocol, e.g. https://some-okapi.somewhere.org */
+let okapiUrl = null;
+
+let okapiTenant = null;
+
+/** categorical logger object */
+// let logger = null;
+
+/** lock to indicate whether a rotation request is already in progress */
+let isRotating = false;
+/** how many times to check the lock before giving up */
+const IS_ROTATING_RETRIES = 100;
+/** how long to wait before rechecking the lock, in milliseconds (100 * 100) === 10 seconds */
+const IS_ROTATING_INTERVAL = 100;
+
+/**
+ * isValidAT
+ * return true if tokenExpiration.atExpires is in the future
+ * @param {object} te tokenExpiration shaped like { atExpires, rtExpires }
+ * @returns boolean
+ */
+export const isValidAT = (te) => {
+  // console.log(`-- (rtr-sw) => at expires ${new Date(te?.atExpires || null).toISOString()}`);
+  return !!(te?.atExpires > Date.now());
+};
+
+/**
+ * isValidRT
+ * return true if tokenExpiration.rtExpires is in the future
+ * @param {object} te tokenExpiration shaped like { atExpires, rtExpires }
+ * @returns boolean
+ */
+export const isValidRT = (te) => {
+  // console.log(`-- (rtr-sw) => rt expires ${new Date(te?.rtExpires || null).toISOString()}`);
+  return !!(te?.rtExpires > Date.now());
+};
+
+/**
+ * messageToClient
+ * Send a message to clients of this service worker
+ * @param {Event} event
+ * @param {*} message
+ * @returns void
+ */
+export const messageToClient = async (event, message) => {
+  // Exit early if we don't have access to the client.
+  // Eg, if it's cross-origin.
+  if (!event.clientId) {
+    // console.log('-- (rtr-sw) PASSTHROUGH: no clientId');
+    return;
+  }
+
+  // Get the client.
+  const client = await self.clients.get(event.clientId);
+  // Exit early if we don't get the client.
+  // Eg, if it closed.
+  if (!client) {
+    // console.log('-- (rtr-sw) PASSTHROUGH: no client');
+    return;
+  }
+
+  // Send a message to the client.
+  // console.log('-- (rtr-sw) => sending', message);
+  client.postMessage({ ...message, source: '@folio/stripes-core' });
+};
+
+/**
+ * rtr
+ * exchange an RT for a new one.
+ * Make a POST request to /authn/refresh, including the current credentials,
+ * and send a TOKEN_EXPIRATION event to clients that includes the new AT/RT
+ * expiration timestamps.
+ * @param {Event} event
+ * @returns Promise
+ * @throws if RTR fails
+ */
+export const rtr = async (event) => {
+  // console.log('-- (rtr-sw) ** RTR ...');
+
+  // if several fetches trigger rtr in a short window, all but the first will
```

Comment:
I'm not completely understanding what problem this would solve. Why would "several fetches trigger rtr in a short window"? Is it that lots of requests can arrive in the filter at once since the browser of course processes 8 or 10 at once in separate threads normally. If this is firing every

The reason I ask is because of the single use feature of the RT. You can only POST a single RT once. If you post it twice (say if two requests send at roughly the same time or in close proximity), mod-at will consider this a "leaked token" and log the user out of all of their devices.

I think this means that the RT should only be sent when it needs to be sent or just lazily in the background, which is at some interval that is less than its TTL. Occasionally it will result in a 401 because it expired.

Reply:
I think what we want is for the service worker to not be too ambitious. He's lazy after all. But his laziness is a virtue. If he does his work well, in his lazy way, he doesn't need to be ambitious. Ambitious would mean that he's trying to solve every problem that comes his way just as soon as these problems arrive. But he's not like that. His reliability is how he makes up for his lack of ambition. He takes care of everything eventually. And because of that he doesn't need to be ambitious.

But why does he care about every request that comes his way then?

Only because just in case, something unforeseen has caused a problem that is outside of his control, he can send the user to the login screen, which is after all, much better than the alternative which would never to be sent anywhere at all.

Reply:
Simultaneous fetches are common, e.g. the ui-users landing page fires several requests to fill in the values on its search form: patron groups, departments, tags, custom fields. The browser handles these in parallel streams, but as you pointed out, the RT is single use, so we need to guard against the situation where an AT has expired and then several requests arrive all at once. We need to let the first one through and force the others to wait. 

As to laziness, this code goes through great pains to be lazy, i.e. to implement just-in-time RTR, passing along requests as-is when it believes the AT is valid. But that's as lazy as we can be: the nature of a fetch event listener is that it hears about every fetch, so we need to handle all the surrounding edge cases: the expiration data in storage said the AT (or RT) was good but it turned out to be bad, here are five simultaneous fetches after a period of inactivity during which the AT expired, etc.

A timer-based approach is simpler, and doesn't need to involve service workers at all, but it limits sessions to the AT's expiration instead of the RT's. If we want sessions that last as long as the RT does, we have to implement JIT renewal as above, and it gets complicated pretty quickly.